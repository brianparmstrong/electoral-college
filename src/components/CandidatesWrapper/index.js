import { useEffect, useState } from 'react';
import { arrayOf, number, shape, string } from 'prop-types';
import { checkForWinner, getWinnerClassName, getWinnerName } from '../../utilities';
import Candidates from '../Candidates';

const CandidatesWrapper = ({
  candidates,
  evPct,
  popVoteTotals,
  pvPct,
  winnerTakeAllTotals
}) => {
  const [wtaWinnerIndex, setWtaWinnerIndex] = useState(checkForWinner(winnerTakeAllTotals));
  const [propWinnerIndex, setPropWinnerIndex] = useState(checkForWinner(popVoteTotals));
  const [wtaWinnerName, setWtaWinnerName] = useState(getWinnerName((wtaWinnerIndex + 1)));
  const [propWinnerName, setPropWinnerName] = useState(getWinnerName((propWinnerIndex + 1)));
  const [wtaWinnerClassName, setWtaWinnerClassName] = useState(getWinnerClassName((wtaWinnerIndex + 1)));
  const [propWinnerClassName, setPropWinnerClassName] = useState(getWinnerClassName((propWinnerIndex + 1)));

  useEffect(() => {
    setPropWinnerIndex(checkForWinner(pvPct));
  }, [ pvPct ]);

  useEffect(() => {
    setWtaWinnerIndex(checkForWinner(evPct));
  }, [ evPct ]);

  useEffect(() => {
    const index = propWinnerIndex + 1;
    const name = getWinnerName(index);
    setPropWinnerName(name);
    setPropWinnerClassName(getWinnerClassName(index));
  }, [ propWinnerIndex ]);

  useEffect(() => {
    const index = wtaWinnerIndex + 1;
    const name = getWinnerName(index);
    setWtaWinnerName(name);
    setWtaWinnerClassName(getWinnerClassName(index));
  }, [ wtaWinnerIndex ]);

  const calculateUnawardedEvsPct = percentages => {
    const awardedEvsPct = percentages.reduce((a, b) => a + b, 0);
    return (100 - awardedEvsPct);
  };

  return (
    <div>
      <h1>Electoral College 2024</h1>
      <div className='votes-counter' id='ev-votes-counter'>
        <h3>Winner-Take-All Electoral Votes Counter</h3>
        <h3 className={`wta-winner ${wtaWinnerClassName} ${wtaWinnerIndex > -1 ? 'show' : 'hide'}`}>
          {`Winner: ${wtaWinnerName}`}
        </h3>
        <div>
          <div className='votes-winner-line'></div>
          <div className='votes-parties'>
            <span className='gop-votes' style={{'width': evPct[0] + '%'}} />
            <span className='lib-votes' style={{'width': evPct[2] + '%'}} />
            <span className='unawarded-votes' style={{'width': calculateUnawardedEvsPct(evPct) + '%'}} />
            <span className='grn-votes' style={{'width': evPct[3] + '%'}} />
            <span className='ind-votes' style={{'width': evPct[4] + '%'}} />
            <span className='dem-votes' style={{'width': evPct[1] + '%'}} />
          </div>
        </div>
      </div>
      <div className='votes-counter' id='pop-votes-counter'>
        <h3>Proportional Electoral Votes Counter</h3>
        <h3 className={`prop-winner ${propWinnerClassName} ${propWinnerIndex > -1 ? 'show' : 'hide'}`}>
          {`Winner: ${propWinnerName}`}
        </h3>
        <div>
          <div className='votes-winner-line'></div>
          <div className='votes-parties'>
            <span className='gop-votes' style={{'width': pvPct[0] + '%'}} />
            <span className='lib-votes' style={{'width': pvPct[2] + '%'}} />
            <span className='unawarded-votes' style={{'width': calculateUnawardedEvsPct(pvPct) + '%'}} />
            <span className='grn-votes' style={{'width': pvPct[3] + '%'}} />
            <span className='ind-votes' style={{'width': pvPct[4] + '%'}} />
            <span className='dem-votes' style={{'width': pvPct[1] + '%'}} />
          </div>
        </div>
      </div>
      <Candidates
        candidatesData={candidates}
        popularVoteTotals={popVoteTotals}
        winnerTakeAllTotals={winnerTakeAllTotals} />
    </div>
  );
};

CandidatesWrapper.propTypes = {
  candidates: arrayOf(shape({
    name: string.isRequired,
    party: string.isRequired
  })).isRequired,
  evPct: arrayOf(number).isRequired,
  popVoteTotals: arrayOf(number).isRequired,
  pvPct: arrayOf(number).isRequired,
  winnerTakeAllTotals: arrayOf(number).isRequired
};

export default CandidatesWrapper;
