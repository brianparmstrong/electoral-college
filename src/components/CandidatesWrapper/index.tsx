import { useEffect, useState } from 'react';
import { IfcCandidatesWrapper } from '../../types';
import {
  checkForWinner,
  getWinnerClassName,
  getWinnerName,
} from '../../utilities';
import Candidates from '../Candidates';

const CandidatesWrapper = (candidatesWrapper: IfcCandidatesWrapper) => {
  const {
    candidateImageSources,
    candidates,
    evPct,
    popVoteTotals,
    pvPct,
    winnerTakeAllTotals,
  } = candidatesWrapper;
  const [wtaWinnerIndex, setWtaWinnerIndex] = useState<number>(
    checkForWinner(winnerTakeAllTotals)
  );
  const [propWinnerIndex, setPropWinnerIndex] = useState<number>(
    checkForWinner(popVoteTotals)
  );
  const [wtaWinnerName, setWtaWinnerName] = useState<string>(
    getWinnerName(String(wtaWinnerIndex + 1))
  );
  const [propWinnerName, setPropWinnerName] = useState<string>(
    getWinnerName(String(propWinnerIndex + 1))
  );
  const [wtaWinnerClassName, setWtaWinnerClassName] = useState<string>(
    getWinnerClassName(String(wtaWinnerIndex + 1))
  );
  const [propWinnerClassName, setPropWinnerClassName] = useState<string>(
    getWinnerClassName(String(propWinnerIndex + 1))
  );

  useEffect(() => {
    setPropWinnerIndex(checkForWinner(pvPct));
  }, [pvPct]);

  useEffect(() => {
    setWtaWinnerIndex(checkForWinner(evPct));
  }, [evPct]);

  useEffect(() => {
    const index = String(propWinnerIndex + 1);
    const name = getWinnerName(index);
    setPropWinnerName(name);
    setPropWinnerClassName(getWinnerClassName(index));
  }, [propWinnerIndex]);

  useEffect(() => {
    const index = String(wtaWinnerIndex + 1);
    const name = getWinnerName(index);
    setWtaWinnerName(name);
    setWtaWinnerClassName(getWinnerClassName(index));
  }, [wtaWinnerIndex]);

  const calculateUnawardedEvsPct = (percentages: Array<number>) => {
    const awardedEvsPct = percentages.reduce((a, b) => a + b, 0);
    return 100 - awardedEvsPct;
  };

  return (
    <div>
      <h1>Electoral College 2024</h1>
      <div className="votes-counter" id="ev-votes-counter">
        <h3>Winner-Take-All Electoral Votes Counter</h3>
        <h3
          className={`wta-winner ${wtaWinnerClassName} ${
            wtaWinnerIndex > -1 ? 'show' : 'hide'
          }`}
        >
          {`Winner: ${wtaWinnerName}`}
        </h3>
        <div>
          <div className="votes-winner-line"></div>
          <div className="votes-parties">
            <span className="gop-votes" style={{ width: evPct[0] + '%' }} />
            <span className="lib-votes" style={{ width: evPct[2] + '%' }} />
            <span
              className="unawarded-votes"
              style={{ width: calculateUnawardedEvsPct(evPct) + '%' }}
            />
            <span className="grn-votes" style={{ width: evPct[3] + '%' }} />
            <span className="ind-votes" style={{ width: evPct[4] + '%' }} />
            <span className="dem-votes" style={{ width: evPct[1] + '%' }} />
          </div>
        </div>
      </div>
      <div className="votes-counter" id="pop-votes-counter">
        <h3>Proportional Electoral Votes Counter</h3>
        <h3
          className={`prop-winner ${propWinnerClassName} ${
            propWinnerIndex > -1 ? 'show' : 'hide'
          }`}
        >
          {`Winner: ${propWinnerName}`}
        </h3>
        <div>
          <div className="votes-winner-line"></div>
          <div className="votes-parties">
            <span className="gop-votes" style={{ width: pvPct[0] + '%' }} />
            <span className="lib-votes" style={{ width: pvPct[2] + '%' }} />
            <span
              className="unawarded-votes"
              style={{ width: calculateUnawardedEvsPct(pvPct) + '%' }}
            />
            <span className="grn-votes" style={{ width: pvPct[3] + '%' }} />
            <span className="ind-votes" style={{ width: pvPct[4] + '%' }} />
            <span className="dem-votes" style={{ width: pvPct[1] + '%' }} />
          </div>
        </div>
      </div>
      <Candidates
        candidateImageSources={candidateImageSources}
        candidatesData={candidates}
        popularVoteTotals={popVoteTotals}
        winnerTakeAllTotals={winnerTakeAllTotals}
      />
    </div>
  );
};

export default CandidatesWrapper;
