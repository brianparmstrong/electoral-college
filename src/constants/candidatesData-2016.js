export const CANDIDATES_DATA = [
  {
    "name": "Trump",
    "party": "gop"
  },
  {
    "name": "Clinton",
    "party": "dem" 
  },
  {
    "name": "Johnson",
    "party": "lib"
  },
  {
    "name": "Stein",
    "party": "grn"
  },
  {
    "name": "McMullin",
    "party": "ind"
  }
];
