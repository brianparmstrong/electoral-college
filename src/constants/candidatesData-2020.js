export const CANDIDATES_DATA = [
  {
    "name": "Trump",
    "party": "gop"
  },
  {
    "name": "Biden",
    "party": "dem" 
  },
  {
    "name": "Jorgensen",
    "party": "lib"
  },
  {
    "name": "Hawkins",
    "party": "grn"
  },
  {
    "name": "West",
    "party": "ind"
  }
];
