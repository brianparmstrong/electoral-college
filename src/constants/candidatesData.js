export const CANDIDATES_DATA = [
  {
    "name": "Rep",
    "party": "gop"
  },
  {
    "name": "Dem",
    "party": "dem" 
  },
  {
    "name": "Lib",
    "party": "lib"
  },
  {
    "name": "Grn",
    "party": "grn"
  },
  {
    "name": "Ind",
    "party": "ind"
  }
];
