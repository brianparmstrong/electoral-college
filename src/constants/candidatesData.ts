import { CandidatesData } from '../types';

export const CANDIDATES_DATA: CandidatesData = [
  {
    name: 'Trump',
    party: 'gop',
  },
  {
    name: 'Biden',
    party: 'dem',
  },
  {
    name: 'Oliver',
    party: 'lib',
  },
  {
    name: 'Stein',
    party: 'grn',
  },
  {
    name: 'Kennedy',
    party: 'ind',
  },
];
