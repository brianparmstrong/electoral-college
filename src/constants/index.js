import { CANDIDATES_DATA } from './candidatesData-2024';
import { PARTY_MAP } from './partyData';
import { STATES_DATA } from './statesData-2020';

export {
  CANDIDATES_DATA,
  PARTY_MAP,
  STATES_DATA
};
