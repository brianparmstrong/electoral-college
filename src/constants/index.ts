import { PARTY_MAP } from './partyData';
import { STATES_DATA } from './statesData';

export { PARTY_MAP, STATES_DATA };
