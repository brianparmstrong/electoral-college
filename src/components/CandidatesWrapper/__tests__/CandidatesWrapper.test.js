import React from 'react';
import renderer from 'react-test-renderer';
import { CANDIDATES_DATA } from '../../../constants';
import CandidatesWrapper from '../';

const MOCK_EV_PCT = [ 15, 14, 0, 0, 0 ];
const MOCK_POP_VOTE_TOTALS = [ 48, 49, 2, 0.5, 0.5 ];
const MOCK_PV_PCT = [ 12.50, 11.50, 2.75, 1.25, 1.00 ];
const MOCK_WINNER_TAKE_ALL_TOTALS = [ 55, 45, 0, 0, 0 ];

it('renders correctly', () => {
  const candidatesWrapper = renderer
    .create(
      <CandidatesWrapper
        candidates={CANDIDATES_DATA}
        evPct={MOCK_EV_PCT}
        popVoteTotals={MOCK_POP_VOTE_TOTALS}
        pvPct={MOCK_PV_PCT}
        winnerTakeAllTotals={MOCK_WINNER_TAKE_ALL_TOTALS} />
    )
    .toJSON();
  expect(candidatesWrapper).toMatchSnapshot();
});
