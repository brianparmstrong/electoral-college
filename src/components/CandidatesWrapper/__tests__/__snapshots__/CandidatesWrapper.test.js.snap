// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1>
    Electoral College 2020
  </h1>
  <div
    className="votes-counter"
    id="ev-votes-counter"
  >
    <h3>
      Winner-Take-All Electoral Votes Counter
    </h3>
    <h3
      className="wta-winner gop show"
    >
      Winner: Trump
    </h3>
    <div>
      <div
        className="votes-winner-line"
      />
      <div
        className="votes-parties"
      >
        <span
          className="gop-votes"
          style={
            Object {
              "width": "15%",
            }
          }
        />
        <span
          className="lib-votes"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <span
          className="unawarded-votes"
          style={
            Object {
              "width": "71%",
            }
          }
        />
        <span
          className="grn-votes"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <span
          className="ind-votes"
          style={
            Object {
              "width": "0%",
            }
          }
        />
        <span
          className="dem-votes"
          style={
            Object {
              "width": "14%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="votes-counter"
    id="pop-votes-counter"
  >
    <h3>
      Proportional Electoral Votes Counter
    </h3>
    <h3
      className="prop-winner  hide"
    >
      Winner: 
    </h3>
    <div>
      <div
        className="votes-winner-line"
      />
      <div
        className="votes-parties"
      >
        <span
          className="gop-votes"
          style={
            Object {
              "width": "12.5%",
            }
          }
        />
        <span
          className="lib-votes"
          style={
            Object {
              "width": "2.75%",
            }
          }
        />
        <span
          className="unawarded-votes"
          style={
            Object {
              "width": "71%",
            }
          }
        />
        <span
          className="grn-votes"
          style={
            Object {
              "width": "1.25%",
            }
          }
        />
        <span
          className="ind-votes"
          style={
            Object {
              "width": "1%",
            }
          }
        />
        <span
          className="dem-votes"
          style={
            Object {
              "width": "11.5%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="candidatesWrapper"
  >
    <div
      className="candidate"
      id="Trump"
    >
      <h2>
        Trump
      </h2>
      <img
        alt="Trump"
        src="trump.jpeg"
      />
      <div
        className="electoralVotesWrapper"
      >
        <p
          className="electoralVotesLabel"
        >
          Winner-Take-All:
        </p>
        <span
          className="electoralVotes"
          id="gop-wta"
        >
          55
        </span>
        <p
          className="electoralVotesLabel"
        >
          Proportional:
        </p>
        <span
          className="electoralVotes"
          id="gop-prop"
        >
          48
        </span>
      </div>
    </div>
    <div
      className="candidate"
      id="Biden"
    >
      <h2>
        Biden
      </h2>
      <img
        alt="Biden"
        src="biden.jpeg"
      />
      <div
        className="electoralVotesWrapper"
      >
        <p
          className="electoralVotesLabel"
        >
          Winner-Take-All:
        </p>
        <span
          className="electoralVotes"
          id="dem-wta"
        >
          45
        </span>
        <p
          className="electoralVotesLabel"
        >
          Proportional:
        </p>
        <span
          className="electoralVotes"
          id="dem-prop"
        >
          49
        </span>
      </div>
    </div>
    <div
      className="candidate"
      id="Jorgensen"
    >
      <h2>
        Jorgensen
      </h2>
      <img
        alt="Jorgensen"
        src="jorgensen.png"
      />
      <div
        className="electoralVotesWrapper"
      >
        <p
          className="electoralVotesLabel"
        >
          Winner-Take-All:
        </p>
        <span
          className="electoralVotes"
          id="lib-wta"
        >
          0
        </span>
        <p
          className="electoralVotesLabel"
        >
          Proportional:
        </p>
        <span
          className="electoralVotes"
          id="lib-prop"
        >
          2
        </span>
      </div>
    </div>
    <div
      className="candidate"
      id="Hawkins"
    >
      <h2>
        Hawkins
      </h2>
      <img
        alt="Hawkins"
        src="hawkins.jpeg"
      />
      <div
        className="electoralVotesWrapper"
      >
        <p
          className="electoralVotesLabel"
        >
          Winner-Take-All:
        </p>
        <span
          className="electoralVotes"
          id="grn-wta"
        >
          0
        </span>
        <p
          className="electoralVotesLabel"
        >
          Proportional:
        </p>
        <span
          className="electoralVotes"
          id="grn-prop"
        >
          0.5
        </span>
      </div>
    </div>
    <div
      className="candidate"
      id="West"
    >
      <h2>
        West
      </h2>
      <img
        alt="West"
        src="west.jpeg"
      />
      <div
        className="electoralVotesWrapper"
      >
        <p
          className="electoralVotesLabel"
        >
          Winner-Take-All:
        </p>
        <span
          className="electoralVotes"
          id="ind-wta"
        >
          0
        </span>
        <p
          className="electoralVotesLabel"
        >
          Proportional:
        </p>
        <span
          className="electoralVotes"
          id="ind-prop"
        >
          0.5
        </span>
      </div>
    </div>
  </div>
</div>
`;
